examples/rust/pipe-ping-pong.rs

```diff
@@ -168,109 +168,164 @@ fn close(libos: &mut LibOS, pipeqd: QDesc) {
     }
 }
 
-//======================================================================================================================
-// server()
-//======================================================================================================================
+// The pipe server.
+pub struct PipeServer {
+    /// Underlying libOS.
+    libos: LibOS,
+    /// Tx pipe qd.
+    tx_pipeqd: Option<QDesc>,
+    /// Rx pipe qd.
+    rx_pipeqd: Option<QDesc>,
+}
 
-fn server(name: &str) -> Result<()> {
-    let libos_name: LibOSName = match LibOSName::from_env() {
-        Ok(libos_name) => libos_name.into(),
-        Err(e) => anyhow::bail!("{:?}", e),
-    };
-    let mut libos: LibOS = match LibOS::new(libos_name) {
-        Ok(libos) => libos,
-        Err(e) => anyhow::bail!("failed to initialize libos: {:?}", e),
-    };
+// Implementation of the pipe server.
+impl PipeServer {
+    pub fn new(libos: LibOS) -> Result<Self> {
+        return Ok(Self {
+            libos,
+            tx_pipeqd: None,
+            rx_pipeqd: None,
+        });
+    }
 
-    // Setup peer.
-    let pipeqd_rx: QDesc = match libos.create_pipe(&format!("{}:rx", name)) {
-        Ok(qd) => qd,
-        Err(e) => anyhow::bail!("failed to open memory queue: {:?}", e),
-    };
+    pub fn run(&mut self, pipe_name: &str) -> Result<()> {
+        // Setup the Rx pipe.
+        self.rx_pipeqd = match self.libos.create_pipe(&format!("{}:rx", pipe_name)) {
+            Ok(qd) => Some(qd),
+            Err(e) => anyhow::bail!("failed to open memory queue: {:?}", e),
+        };
 
-    // Setup peer.
-    let pipeqd_tx: QDesc = match libos.create_pipe(&format!("{}:tx", name)) {
-        Ok(qd) => qd,
-        Err(e) => {
-            // Clean up open pipes on error.
-            close(&mut libos, pipeqd_rx);
-            anyhow::bail!("failed to open memory queue: {:?}", e)
-        },
-    };
+        // Setup the Tx pipe.
+        self.tx_pipeqd = match self.libos.create_pipe(&format!("{}:tx", pipe_name)) {
+            Ok(qd) => Some(qd),
+            Err(e) => {
+                if let Some(rx_pipeqd) = self.rx_pipeqd {
+                    close(&mut self.libos, rx_pipeqd);
```

Comment:
`close()` may fail. Could you capture the return value and warn on fail?

Reply:
The close() function captures the return value and prints/warns.

Reply:
I see. Thus if `close` operates on a property of the given structure, close should be an associated function as well.

Reply:
Done.